h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostUtilsTest.java

```diff
@@ -314,6 +309,42 @@ public void testSparsematrixInit_dimensions_test() throws XGBoostError {
       }
     }
 
+    @Test
+    public void testSumChunksLength() {
+      try {
+        Scope.enter();
+
+        final Vec dataVec = Vec.makeCon(0d, 1000);
+        dataVec.set(0, 1);
+        final int[] localChunkIds = VecUtils.getLocalChunkIds(dataVec);
+        int[] localChunksLengths = new int[localChunkIds.length];
+
+        // Without weights, all rows should be counted
+        long nonZeroRows = XGBoostUtils.sumChunksLength(localChunkIds, dataVec, Optional.empty(), localChunksLengths);
+        assertEquals(dataVec.length(), nonZeroRows);
+
+        // With all weights as 1, all rows should be counted
+        final Vec allWeightsOne = Vec.makeCon(1d, 1000);
+        localChunksLengths = new int[localChunksLengths.length];
+        long weightedNonZeroRows = XGBoostUtils.sumChunksLength(localChunkIds, dataVec, Optional.of(allWeightsOne), localChunksLengths);
+        assertEquals(dataVec.length(), weightedNonZeroRows);
+        
+        // Mixed weights - only the non-zero weight rows should be counted
+        final Vec mixedWeights = Vec.makeCon(0d, 1000);
```

Comment:
This part tests the root cause of this problem:

All chunks with one exception have trivial implementation of `nextNz` - just give the next element. Might not always be correct, but that's how it works.

```java
public int nextNZ(int rid){ return rid + 1;}
```

The problem was simple - the original code, while seemingly having bad logic, worked in most cases, as it inspected all chunks one by one anyway, due to the implementation of `nextNz` in most chunks.

The problem arises with `CXIChunk` - the zero-sparse variant. It's `nextNz` method is implemented as follows:

```java
  @Override public final int nextNZ(int i){
    int x = findOffset(i);
    if(x < 0) x = -x-1-_elem_sz;
    _previousOffset = x += _elem_sz;
    return getId(x);
  }
```

This made the original logic return nonsensical numbers, as it actually skipped values.

I was unable to reproduce this behavior every time with the given test (see https://0xdata.atlassian.net/browse/PUBDEV-7819), probably due to different chunking of the dataset on my machine. Anyway, this unit test simulates the bad behavior all the time and actually tests the behavior is correct.